stdlib/fun.mli

```diff
@@ -67,3 +69,199 @@ exception Finally_raised of exn
+
+
+{3:hid {{!val:id}id}}
+
+{!val:List.init} with the index itself
+{[
+# List.init 3 Fun.id;;
+- : int list = [0; 1; 2]
+]}
+
+Using {!val:List.filter_map} on an [int option list] to filter out
+{{!const:Option.t.None}[None]} elements
+{[
+# List.filter_map Fun.id [None; Some 2; Some 3; None; Some 5];;
+- : int list = [2; 3; 5]
+]}
+
+Using {!val:Float.Array.map_from_array} to flatten a [float array]
+{[
+# let to_flat arr = Float.Array.map_from_array Fun.id arr;;
+val to_flat : float array -> Float.Array.t
+]}
+
+Dispatching functions of type [foo -> foo] conditionally is another place where
+[id] may be useful. Consider a function which either uses {!val:String.map} on
+a path to convert unix-style slashes to windows-style backslashes or leaves it
+as it is depending on the {!val:Sys.win32} value
+{[
+if Sys.win32 then String.map (function '/' -> '\\' | c -> c) else Fun.id
+]}
+
+And for more advanced uses, where we may build up closures, [id] is often used
+for the base-case as a no-op. Consider a function which chains a list of unary
+functions:
+{[
+let rec chain = function
+  | [] -> Fun.id
+  | f :: fs -> fun x -> f (chain fs x)
+]}
+
+
+{3:hconst {{!val:const}const}}
+
+{!val:List.init} a list of zeros
+{[
+# List.init 3 (Fun.const 0);;
+- : int list = [0; 0; 0]
+]}
+
+Note that applying [const (...)] evaluates the expression [(...)] once, and
+returns a function that only has the result of this evaluation. To demonstrate
+this, consider if [(...)] was a call to {!val:Random.bool}[()]:
+
+[List.init n (Fun.const (Random.bool()))] for any [n > 0] will have {e exactly
+two} possible outcomes, [[true; true; ...; true]] or
+[[false; false; ...; false]], whereas [List.init n (fun _ -> Random.bool())]
+will have 2{^n} possible outcomes, because the randomness effect is performed
+with every element.
+
+For more real-world uses, consider {!val:String.spellcheck} with a constant max
+distance of 2
+{[
+# let spellcheck known_words word =
+    let dict_iter yield = List.iter yield known_words in
+    String.spellcheck ~max_dist:(Fun.const 2) dict_iter word
+  ;;
+val spellcheck : string list -> string -> string list
+]}
+
+
+{3:hflip {{!val:flip}flip}}
+
+Use [flip] to reverse the comparator passed to {!val:List.sort}, resulting in a
+reversed sorting
+{[
+# List.sort (Fun.flip Int.compare) [5; 3; 9; 0; 1; 6; 8];;
+- : int list = [9; 8; 6; 5; 3; 1; 0]
+]}
+
+Reverse a list by accumulating a new list using {!val:List.fold_left}, which
+expects the accumulator to be the first argument of the function passed to it.
+We pass {!val:List.cons} which has the list as the second argument, so [flip]
+is useful here
+{[
+# List.fold_left (Fun.flip List.cons) [] [1; 2; 3];;
+- : int list = [3; 2; 1]
+]}
+
+Using {!val:List.find_all} and a flipped {!val:List.mem} to get an intersection
+of two lists (with order and duplicates from the second)
+{[
+# List.find_all (Fun.flip List.mem [2; 3; 5]) [0; 3; 3; 2; 4; 6; 8]
+- : int list = [3; 3; 2]
```

Comment:
I don't really like this example and the one below. They use the form `flip f x` to partially-apply the second argument of `f`, which I think is too clever, I would rather write an explicit function in this case.

Reply:
That was what I was going for, `flip` with the second argument partially applied. I don't mind removing it.
The rationale was that it was common enough (you see it in Set, Map, List, Stack, Queue, ... wherever you want to make datafirst become datalast or the opposite)